### PR TITLE
[X] Reuse baseclass namescope if it exists

### LINKED ISSUE #7097: Xamarin.Forms.Xaml.XamlParseException with Bindings Reference in XF 4.2 Pre 3
### Description
In xamarin forms 4.2 pre 3 I get this exception:
> Xamarin.Forms.Xaml.XamlParseException: 'Position 60:67. Can not find the object referenced by OrderList'

the exception is related to the binding to a reference. Like this line:
```xaml
     ColumnDefinition Width="{Binding Source={x:Reference Name=OrderList}, Path=BindingContext.IsLocationDeleteVisible, Converter={StaticResource VariableGridLengthConverter}, ConverterParameter=OrderColumn1}"
```
                                           
Commenting this line out I get the exception in other lines containing a reference.
It is working with XF 4.2 pre 2 and earlier (3.x).

### Basic Information

- Version with issue: https://github.com/xamarin/Xamarin.Forms/releases/tag/beta-4.2.0-pre3.1
- Last known good version: https://github.com/xamarin/Xamarin.Forms/releases/tag/beta-4.2.0-pre2
- IDE:
Microsoft Visual Studio Professional 2019
Version 16.2.1
VisualStudio.16.Release/16.2.1+29201.188
Microsoft .NET Framework
Version 4.7.03056

- Platform Target Frameworks: 
  - Android: 9.0 (Pie)
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

### Reproduction Link
Is part of a big project. Which is not applied here.

### LINKED ISSUE #7286: [Bug] All XAML-defined elements null in Page Constructors with 4.2.0.709249
### Description
After updating to the latest Version of Xamarin Forms (4.2.0.709249) on Android and iOS, my App throws an exception at start. Investigating the issue showed that in the constructor call of my initial Page, all XAML-defined elements are null. The issue was fixed by downgrading the Xamarin Forms Nuget package.
This happens with multiple pages in my project, I have tried starting from several to exclude a mistake in the specific page. 

### Steps to Reproduce

1. Start Application, Load a new App. (either iOS or Android)
2. Set a MainPage in App.xaml.cs in the overridden OnStart.
3. In Constructor of set MainPage, make a call to an element specified in XAML of this page (e.g. set a Label).
4. Label is null, as any other element from the XAML.

### Expected Behavior
App launches as usual.

### Actual Behavior
Exception thrown in the constructor of the first loaded page when trying to access an element of the XAML code.

### Basic Information

- Version with issue: 4.2.0.709249
- Last known good version: 4.1.0.673156
- IDE: Visual Studio for Mac
- Platform Target Frameworks: 
  - Shared Code: .NET Standard 2.0, Library
  - iOS:  Deployment Target 9.0
  - Android: 9.0 Pie
- Affected Devices: all

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

If a xaml control derives from another xaml control, thy both should share the same
NameScope. This add a check if the xaml root already has a NameScope (set by the base)
and reuses it.

This fixes 2 regressions introduced by #7040 and some other unreported edge cases.

### Issues Resolved ### 

- fixes #7097
- fixes #7286

### API Changes ###

 
 None

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [x] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Build.Tasks/SetNamescopesAndRegisterNamesVisitor.cs [hunk 1]
```diff
@@ -77,33 +74,38 @@ public void Visit(ListNode node, INode parentNode)
...
+			=> parentNode is IElementNode parentElement && parentElement.Properties.TryGetValue(XmlName.xName, out INode xNameNode) && xNameNode == node;
+
+		VariableDefinition GetOrCreateNameScope(ElementNode node)
 		{
-			var parentElement = parentNode as IElementNode;
-			INode xNameNode;
-			if (parentElement != null && parentElement.Properties.TryGetValue(XmlName.xName, out xNameNode) && xNameNode == node)
-				return true;
-			return false;
+			var module = Context.Body.Method.Module;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Build.Tasks/SetNamescopesAndRegisterNamesVisitor.cs]: this is the core of the change for XamlC

File: Xamarin.Forms.Xaml/CreateValuesVisitor.cs [hunk 2]
```diff
@@ -157,16 +157,18 @@ public void Visit(RootNode node, INode parentNode)
 			var rnode = (XamlLoader.RuntimeRootNode)node;
 			Values[node] = rnode.Root;
 			Context.Types[node] = rnode.Root.GetType();
-			var bindableRoot = rnode.Root as BindableObject;
-			if (bindableRoot != null)
-				NameScope.SetNameScope(bindableRoot, node.Namescope);
+			if (rnode.Root is BindableObject bindable) {
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Xaml/CreateValuesVisitor.cs]: and this is the part for non-XamlC

File: Xamarin.Forms.Build.Tasks/DebugXamlCTask.cs [hunk 3]
```diff
@@ -113,8 +113,16 @@ public override bool Execute(out IList<Exception> thrownExceptions)
 						var br2 = Instruction.Create(OpCodes.Ldarg_0);
 						var ret = Instruction.Create(OpCodes.Ret);
 						il.Emit(OpCodes.Ldarg_0);
-						var baseCtor = module.ImportReference(typeDef.BaseType.Resolve().GetConstructors().First(c => c.HasParameters == false));
-						baseCtor = module.ImportReference(baseCtor.ResolveGenericParameters(typeDef.BaseType, module));
+						MethodReference baseCtor;
+						if (typeDef.BaseType.Resolve().GetConstructors().FirstOrDefault(c => c.HasParameters && c.Parameters.Count == 1 && c.Parameters[0].Name == "useCompiledXaml") != null) {
+							baseCtor = module.ImportReference(typeDef.BaseType.Resolve().GetConstructors().First(c => c.HasParameters && c.Parameters.Count == 1 && c.Parameters[0].Name == "useCompiledXaml"));
```
Comment (hartez) [inline Xamarin.Forms.Build.Tasks/DebugXamlCTask.cs]: Can we simplify this so we don't have to repeat this Linq query for the constructor?

Comment (StephaneDelcroix) [reply Xamarin.Forms.Build.Tasks/DebugXamlCTask.cs]: not a critical path (this code is only used when compiling Unit Tests), but you're right

Comment (StephaneDelcroix) [reply Xamarin.Forms.Build.Tasks/DebugXamlCTask.cs]: fixed

**Comment by devology-rob** [comment]
I don't know if this is related, but my app works in 4.2.0.709249 but breaks in 4.2.0.778463 (service release 1) due to an unhandled exception...

```
System.ArgumentException: An Element with the same key already exists in NameScope
```

The top-line of the stack refers to

```
INameScope.RegisterName (line 21)
```

I can't see anything else in the release notes that seems faintly related, but I don't know the Xamarin.Forms codebase.

Please note that in our implementation we call InitializeComponent from our page constructor, but it's also invoked in response to an event due to a workaround where some bindings never re-rendered properly and this heavyweight workaround 'just worked' until 4.2.0.778463

**Comment by samhouts** [comment]
@devology-rob Yes, this is likely the same error. Please try the service release 2 that was released today. Thanks!

**Comment by devology-rob** [comment]
Thanks @samhouts for the rapid reply.

Xamarin.Forms 4.2.0.815419 (4.2.0 Service Release 2) is what I updated my Android and common project nugets to first, but the app failed at runtime with the aforementioned error, so I rolled back to Release 1 then finally 709249 which worked.  

That's why I looked in SR1 to find out what changed, that's where I found this PR in the hope it justifies the crash.

**Comment by samhouts** [comment]
@devology-rob Okay, please file a new issue. Thanks!

**Pull request merged**